### PR TITLE
Debugging login flow

### PR DESCRIPTION
**Pull request by viktor-yakubiv**
@Joozty I face the problem with URLs here and would be happy to hear your advice.

Store becomes first source of truth. It initialises from route first time and then all routes should reflect state of store.

Currently we have 2 redirect pages. `/` (root) index and `/data-providers` index. I think we have to remove these pages and handle routes somehow through store. The problem is that we have to handle other routes somehow through store what makes Next's implicit routing completely useless.

Do you have any opinion?

### CODE REVIEW

**Comment by Joozty** [comment]
> Store becomes first source of truth. It initialises from route first time and then all routes should reflect state of store.

It works like that now there is no need to do any change, is there? Redirection happens on client-side. `componentDidMount` in _app.jsx is called only once on the client-side. We also [check whether root store has been initialized or not](https://github.com/oacore/dashboard/blob/master/store/index.jsx#L8).

File: store/user.js [hunk 1]
```diff
@@ -1,81 +1,36 @@
...
-        {},
-        {},
-        true
-      )
-      this.repositories = data.organisation.repositories
-      this.selectedRepositoryId = this.repositories[0].id
-    } catch (e) {
-      if (e.status === 404) throw e
-    }
+    extendObservable(this, data)
```
Comment (Joozty) [inline store/user.js]: 👍

**Comment by viktor-yakubiv** [comment]
@Joozty I managed to make this work. I will prettify commits in next hour or 2.

**Comment by viktor-yakubiv** [comment]
@Joozty it's kind of final stage. Please, read commit messages.

File: store/root.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    )
+
+    if (dataProvider == null) {
+      throw new Error(
+        `User ${this.user.id} does not have access to DataProvider ${dataProviderId}.`
+      )
+    }
+
+    this.dataProvider = dataProvider
+    this.changeActivity('overview')
```
Comment (Joozty) [inline store/root.js]: IMHO user should stay on the page where they were.

Comment (viktor-yakubiv) [reply store/root.js]: No. If data provider changes users starts from home. Secondary navigation does not exist when primary navigates.

File: pages/_app/index.jsx [hunk 3]
```diff
@@ -21,12 +26,40 @@ class App extends NextApp {
...
+
+    // Assumes store object is always the same
+    this.store = store
+    this.disposeRouter = autorun(this.reflectStoreToRoute.bind(this))
+
+    try {
+      const { dataProvider, activity } = new Route(window.location.pathname)
+      await store.init(dataProvider, activity)
+      this.setState({ isAuthorized: true })
+    } catch (unauthorizedError) {
```
Comment (Joozty) [inline pages/_app/index.jsx]: This error never will never happen I think.

File: pages/_app/index.jsx [hunk 4]
```diff
@@ -21,12 +26,40 @@ class App extends NextApp {
...
+    // TODO: This probably should be under condition
+    router.push(route.href, route.as)
+  }
+
+  async componentDidMount() {
+    const store = initStore()
+
+    // Assumes store object is always the same
+    this.store = store
+    this.disposeRouter = autorun(this.reflectStoreToRoute.bind(this))
```
Comment (Joozty) [inline pages/_app/index.jsx]: I always considered using autorun outside of the store like a bad practice but I don't know why. :D

File: pages/_app/index.jsx (hunk 3 above)
Comment (viktor-yakubiv) [reply pages/_app/index.jsx]: Yes but it should. If user is not authorised, cookie changes or something, we get 500. It's more API problem, we can expect 401.

File: store/deposit-dates.js [hunk 5]
```diff
@@ -46,9 +42,9 @@ class DepositDates {
     return Math.round(level * 100) / 100
   }
 
-  retrieveDepositDates = async (pageNumber, searchTerm, columnOrder) => {
+  async retrieveDepositDates(pageNumber, searchTerm, columnOrder) {
     const order = getOrder(columnOrder)
-    const key = `${this.rootStore.dataProvider}-${pageNumber}-${searchTerm}-${order}`
+    const key = `${pageNumber}-${searchTerm}-${order}`
```
Comment (Joozty) [inline store/deposit-dates.js]: I don't think this was a good move. If you switch dataProvider you will see wrong data.

File: pages/_app/index.jsx (hunk 4 above)
Comment (viktor-yakubiv) [reply pages/_app/index.jsx]: In my opinion, `autorun` is a simple `observer`. Comparing to `computed` that runs on request, `autorun` runs always.

Maybe you have a point. I will think about it more.

File: store/deposit-dates.js (hunk 5 above)
Comment (viktor-yakubiv) [reply store/deposit-dates.js]: No. I recreate child stores when data provider changes.

File: pages/_app/index.jsx (hunk 3 above)
Comment (Joozty) [reply pages/_app/index.jsx]: Sorry I didn't notice this return statement.

https://github.com/oacore/dashboard/blob/login-debugging/api/index.js#L34

Comment (viktor-yakubiv) [reply pages/_app/index.jsx]: Honestly, I forgot to remove that part and throw error instead.

Comment (viktor-yakubiv) [reply pages/_app/index.jsx]: I've just did it.

File: store/root.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  @observable works = null
+
+  @observable depositDates = null
+
+  @computed
+  get dataProviders() {
+    return this.user.dataProviders
+  }
+
+  @action async init(dataProviderId, activityPath) {
```
Comment (Joozty) [inline store/root.js]: I was thinking whether this should be marked as an action or not. Do you have any opinion about it?

[From docs](https://mobx.js.org/refguide/action.html):

> Actions should only, and always, be used on functions that modify state. Functions that just perform look-ups, filters etc should not be marked as actions; to allow MobX to track their invocations.

I haven't found a big advantage in using actions yet. I guess it's good for debugging store state with [Mobx Devtools](https://github.com/mobxjs/mobx-devtools)

Comment (viktor-yakubiv) [reply store/root.js]: I marked with `@action` each function that modifies the state according to the doc. If function calls the function that modifies the state, it's an action too. I am not sure about the second but agree to the first one.

**Pull request merged**